tpm2tools/keys.go

```diff
@@ -133,3 +135,160 @@ func (k *Key) PublicKey() crypto.PublicKey {
+	auth, err := getPCRSessionAuth(rw, pcrs)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get pcr session auth: %v", err)
+	}
+
+	return tpm2.Seal(rw, k.Handle(), "", "", auth, sensitive)
+}
+
+// Unseal takes a private/public pair of buffers and attempts to reverse the
+// sealing process under the owner hierarchy using the SHA256 versions of the
+// provided PCRs.
+// The Key k is used as the parent key.
+func (k *Key) Unseal(rw io.ReadWriter, pcrs []int, priv, pub []byte) ([]byte, error) {
+	session, err := createPCRSession(rw, pcrs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unseal: %v", err)
+	}
+	defer tpm2.FlushContext(rw, session)
+
+	sealed, _, err := tpm2.Load(
+		rw,
+		k.Handle(),
+		/*parentPassword=*/ "",
+		pub,
+		priv)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load sealed object: %v", err)
+	}
+	defer tpm2.FlushContext(rw, sealed)
+
+	return tpm2.UnsealWithSession(rw, session, sealed, "")
+}
+
+// Reseal unwraps a secret and rewraps it under the auth value that would be
+// produced by the PCR state in pcrs. Similar to seal and unseal, this acts on
+// the SHA256 PCRs and uses the ownerhierarchy.
+// The Key k is used as the parent key.
+func (k *Key) Reseal(rw io.ReadWriter, pcrs map[int][]byte, priv, pub []byte) ([]byte, []byte, error) {
+	pcrNums := make([]int, 0, len(pcrs))
+	for key := range pcrs {
+		pcrNums = append(pcrNums, key)
+	}
+
+	sensitive, err := k.Unseal(rw, pcrNums, priv, pub)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to unseal: %v", err)
+	}
+
+	auth, err := computePCRSessionAuth(pcrs)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to compute pcr session auth: %v", err)
+	}
+
+	return tpm2.Seal(rw, k.Handle(), "", "", auth, sensitive)
+}
+
+type tpmsPCRSelection struct {
+	Hash tpm2.Algorithm
+	Size byte
+	PCRs tpmutil.RawBytes
+}
+
+type sessionSummary struct {
+	OldDigest      tpmutil.RawBytes
+	CmdIDPolicyPCR uint32
+	NumPcrSels     uint32
+	Sel            tpmsPCRSelection
+	PcrDigest      tpmutil.RawBytes
+}
+
+func computePCRSessionAuth(pcrs map[int][]byte) ([]byte, error) {
+	var pcrBits [3]byte
+	for pcr := range pcrs {
+		byteNum := pcr / 8
+		bytePos := byte(1 << byte(pcr%8))
+		pcrBits[byteNum] |= bytePos
+	}
+	pcrDigest := digestPCRList(pcrs)
+
+	summary := sessionSummary{
+		OldDigest:      make([]byte, sha256.Size),
+		CmdIDPolicyPCR: uint32(0x0000017f),
+		NumPcrSels:     uint32(1),
+		Sel: tpmsPCRSelection{
+			Hash: tpm2.AlgSHA256,
+			Size: uint8(3),
+			PCRs: pcrBits[:],
+		},
+		PcrDigest: pcrDigest,
+	}
+	b, err := tpmutil.Pack(summary)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pack for hashing: %v ", err)
+	}
+
+	digest := sha256.Sum256(b)
+	return digest[:], nil
+}
+
+func digestPCRList(pcrs map[int][]byte) []byte {
+	hash := crypto.SHA256.New()
+	for i := 0; i < 24; i++ {
+		if pcrValue, exists := pcrs[i]; exists {
+			binary.Write(hash, binary.BigEndian, pcrValue)
```

Comment:
hash.Write(pcrValue)

It's a byte array so endianness does not matter

Reply:
Done.